Plutarch/Builtin.hs

```diff
@@ -71,6 +76,20 @@ ptailBuiltin = phoistAcyclic $ pforce $ punsafeBuiltin PLC.TailList
 pnullBuiltin :: Term s (PBuiltinList a :--> PBool)
 pnullBuiltin = phoistAcyclic $ pforce $ punsafeBuiltin PLC.NullList
 
+{- | 
+  Map type used for Plutus `Data`'s Map constructor.
+
+  Note that the Plutus API doesn't use this most of the time,
+  instead encoding as a List of Tuple constructors.
+
+  Not to be confused with `PlutusTx.AssocMap.Map` / `PMap`
+
+-}
+type PBuiltinMap a b = (PBuiltinList (PBuiltinPair (PAsData a) (PAsData b)))
+
+-- | List type used for Plutus `Data`'s List constructor.
+type PList a  = PBuiltinList (PAsData a)
```

Comment:
I worry about whether or not it's redundant to write something like `PAsData PTxInfo` (in the case of `PList PTxInfo`). After all `PTxInfo` *is* `Data`, as opposed to, say `PPubKeyHash` - which does require this treatment.

Reply:
The internal representation of `TxInfo` is an implementation detail.

Reply:
There's a choice I've made here for consistency -
all of the types are assumed to be 'unwrapped' by default from the Data constructor. 

As @L-as says, the internal repr is an implementation detail, the repr of `PAsData` can be arbitrarily.